### PR TITLE
[libunwind] Add v1.8.1

### PR DESCRIPTION
**Pull request by ararslan**
List of changes: https://github.com/libunwind/libunwind/discussions/698

### CODE REVIEW

**Comment by giordano** [comment]
https://buildkite.com/julialang/yggdrasil/builds/7797#018d136e-5e99-499b-a460-7d62227df9f9/620-2148
```
[18:01:19] aarch64/Gos-linux.c: In function ‘_Uaarch64_local_resume’:
[18:01:19] aarch64/Gos-linux.c:147:1: error: x29 cannot be used in asm here
[18:01:19]  }
[18:01:19]  ^
[18:01:19] aarch64/Gos-linux.c:41:7: error: ‘asm’ operand has impossible constraints
[18:01:19]        __asm__ __volatile__ (
[18:01:19]        ^
```
May need a newer gcc to compile this asm, I've seen similar assembler error before on aaarch64

File: L/LibUnwind/LibUnwind@1.8.0/build_tarballs.jl [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    BuildDependency("XZ_jll"),
+    Dependency("Zlib_jll"),
+    BuildDependency("LLVMCompilerRT_jll"; platforms=[Platform("x86_64", "linux"; sanitize="memory")]),
+]
+
+# Build the tarballs. Note that libunwind started using `stdatomic.h`, which is only
+# available with GCC version 4.9 or later, so we need to set a higher preferred version
+# than the default.
+build_tarballs(ARGS, name, version, sources, script, platforms, products, dependencies;
+               julia_compat="1.10", preferred_gcc_version=v"6")
```
Comment (ararslan) [inline L/LibUnwind/LibUnwind@1.8.0/build_tarballs.jl:78]:
```suggestion
               julia_compat="1.10", preferred_gcc_version=v"12")
```
Semi-random guess (GCC 12 was mentioned in https://github.com/libunwind/libunwind/pull/503)

File: L/LibUnwind/LibUnwind@1.8.1/bundled/patches/libunwind-configure-static-lzma.patch [hunk 2]
```diff
@@ -0,0 +1,20 @@
+--- configure.orig	2023-06-04 05:19:04
++++ configure	2023-06-07 08:35:11
+@@ -18117,7 +18117,7 @@
+   $as_echo_n "(cached) " >&6
+ else
+   ac_check_lib_save_LIBS=$LIBS
+-LIBS="-llzma  $LIBS"
++LIBS="-L/workspace/destdir/lib -l:liblzma.a $LIBS"
```
Comment (giordano) [inline L/LibUnwind/LibUnwind@1.8.1/bundled/patches/libunwind-configure-static-lzma.patch:8]: Can't this be achieved by setting some vars, rather then patching the configure script?
```suggestion
+LIBS="-L${libdir} -l:liblzma.a $LIBS"
```

File: L/LibUnwind/LibUnwind@1.8.1/bundled/patches/libunwind-configure-static-lzma.patch [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+ cat confdefs.h - <<_ACEOF >conftest.$ac_ext
+ /* end confdefs.h.  */
+ 
+@@ -18148,7 +18148,7 @@
+ { $as_echo "$as_me:${as_lineno-$LINENO}: result: $ac_cv_lib_lzma_lzma_mf_is_supported" >&5
+ $as_echo "$ac_cv_lib_lzma_lzma_mf_is_supported" >&6; }
+ if test "x$ac_cv_lib_lzma_lzma_mf_is_supported" = xyes; then :
+-  LIBLZMA=-llzma
++  LIBLZMA="-L/workspace/destdir/lib -l:liblzma.a"
```
Comment (giordano) [inline L/LibUnwind/LibUnwind@1.8.1/bundled/patches/libunwind-configure-static-lzma.patch:17]:
```suggestion
+  LIBLZMA="-L${libdir} -l:liblzma.a"
```

File: L/LibUnwind/LibUnwind@1.8.1/build_tarballs.jl [hunk 4]
```diff
@@ -0,0 +1,81 @@
+using BinaryBuilder
```
Comment (giordano) [inline L/LibUnwind/LibUnwind@1.8.1/build_tarballs.jl:1]:
```suggestion
using BinaryBuilder, Pkg
using BinaryBuilderBase: sanitize
```

File: L/LibUnwind/LibUnwind@1.8.1/build_tarballs.jl [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+# The products that we will ensure are always built
+products = [
+    LibraryProduct("libunwind", :libunwind),
+]
+
+# Dependencies that must be installed before this package can be built
+dependencies = [
+    BuildDependency("XZ_jll"),
+    Dependency("Zlib_jll"),
+    BuildDependency("LLVMCompilerRT_jll"; platforms=[Platform("x86_64", "linux"; sanitize="memory")]),
```
Comment (giordano) [inline L/LibUnwind/LibUnwind@1.8.1/build_tarballs.jl:74]:
```suggestion
    BuildDependency(PackageSpec(name="LLVMCompilerRT_jll", uuid="4e17d02c-6bf5-513e-be62-445f41c75a11", version=llvm_version); platforms=filter(p -> sanitize(p)=="memory", platforms)),
```

File: L/LibUnwind/LibUnwind@1.8.1/build_tarballs.jl [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+# platforms are passed in on the command line.  libunwind is only used
+# on Linux or FreeBSD (e.g. ELF systems)
+platforms = filter(p -> Sys.islinux(p) || Sys.isfreebsd(p), supported_platforms())
+push!(platforms, Platform("x86_64", "linux"; sanitize="memory"))
+
+# The products that we will ensure are always built
+products = [
+    LibraryProduct("libunwind", :libunwind),
+]
+
```
Comment (giordano) [inline L/LibUnwind/LibUnwind@1.8.1/build_tarballs.jl:69]:
```suggestion

llvm_version = v"13.0.1"

```

File: L/LibUnwind/LibUnwind@1.8.1/build_tarballs.jl [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+    BuildDependency("XZ_jll"),
+    Dependency("Zlib_jll"),
+    BuildDependency("LLVMCompilerRT_jll"; platforms=[Platform("x86_64", "linux"; sanitize="memory")]),
+]
+
+# Build the tarballs. Note that libunwind started using `stdatomic.h`, which is only
+# available with GCC version 4.9 or later, so we need to set a higher preferred version
+# than the default.
+build_tarballs(ARGS, name, version, sources, script, platforms, products, dependencies;
+               julia_compat="1.10", preferred_gcc_version=v"6")
```
Comment (giordano) [inline L/LibUnwind/LibUnwind@1.8.1/build_tarballs.jl:81]:
```suggestion
               julia_compat="1.10", preferred_gcc_version=v"6", preferred_llvm_version=llvm_version)
```

**Comment by ararslan** [comment]
Ah, this is apparently a known issue: https://github.com/libunwind/libunwind/issues/742

**Comment by ararslan** [comment]
I've included https://github.com/libunwind/libunwind/pull/748 as a patch, which fixes building on Linux AArch64. An alternative approach would have been to include a patch that reverts https://github.com/libunwind/libunwind/pull/503, which is the approach that the .NET runtime took when upgrading to libunwind 1.8.0 (https://github.com/dotnet/runtime/pull/96969).

File: L/LibUnwind/LibUnwind@1.8.1/build_tarballs.jl [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+# on Linux or FreeBSD (e.g. ELF systems)
+platforms = filter(p -> Sys.islinux(p) || Sys.isfreebsd(p), supported_platforms())
+push!(platforms, Platform("x86_64", "linux"; sanitize="memory"))
+
+# The products that we will ensure are always built
+products = [
+    LibraryProduct("libunwind", :libunwind),
+]
+
+llvm_version = v"13.0.1"
```
Comment (giordano) [inline L/LibUnwind/LibUnwind@1.8.1/build_tarballs.jl:71]:
```suggestion
llvm_version = v"16.0.6"
```

**Pull request merged**